tracing-core/src/callsite.rs

```diff
@@ -169,3 +177,169 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+unsafe impl Send for LinkedList {}
+unsafe impl Sync for LinkedList {}
+
+/// Represents a registration within the callsite cache.
+///
+/// This type enables the intrusive aspect of the callsite cache by
+/// avoiding allocations and letting the callsite's themselves store
+/// the next callsite in the cache.
+pub struct Registration {
+    #[doc(hidden)]
+    pub callsite: &'static dyn Callsite,
+    #[doc(hidden)]
+    pub next: AtomicPtr<Registration>,
+}
+
+impl LinkedList {
+    fn new() -> Self {
+        LinkedList {
+            head: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+
+    fn for_each(&self, mut f: impl FnMut(&'static dyn Callsite)) {
+        let mut head = self.head.load(Ordering::SeqCst);
+
+        loop {
+            if !head.is_null() {
+                let reg = unsafe { &*head };
+                f(reg.callsite);
+
+                head = reg.next.load(Ordering::SeqCst);
+            } else {
+                break;
+            }
+        }
```

Comment:
TIOLI: rather than `loop`/`break`, this could just be
```suggestion
        while !head.is_null() {
            let reg = unsafe { &*head };
            f(reg.callsite);

            head = reg.next.load(Ordering::SeqCst);
        }
```